xLights/FSEQFile.cpp

```diff
@@ -404,20 +404,72 @@ void FSEQFile::preload(uint64_t pos, uint64_t size) {
 #endif
 }
 
-void FSEQFile::parseVariableHeaders(const std::vector<uint8_t> &header, int start) {
-    while (start < header.size() - 5) {
-        int len = read2ByteUInt(&header[start]);
-        if (len) {
+inline bool isRecongizedVariableHeader(uint8_t a, uint8_t b) {
+    // mf - media filename
+    // sp - sequence producer
+    // see https://github.com/FalconChristmas/fpp/blob/master/docs/FSEQ_Sequence_File_Format.txt#L48 for more information
+    return (a == 'm' && b == 'f') || (a == 's' && b == 'p');
+}
+
+const int FSEQ_VARIABLE_HEADER_SIZE = 4;
+
+void FSEQFile::parseVariableHeaders(const std::vector<uint8_t> &header, int readIndex) {
+    const int VariableLengthSize = 2, VariableCodeSize = 2;
+
+    // when encoding, the header size is rounded to the nearest multiple of 4
+    // this comparison ensures that there is enough bytes left to at least constitute a 2 byte length + 2 byte code
+    while (readIndex + FSEQ_VARIABLE_HEADER_SIZE < header.size()) {
+        const int dataLength = read2ByteUInt(&header[readIndex]) - FSEQ_VARIABLE_HEADER_SIZE;
+
+        readIndex += VariableLengthSize;
+
+        if (dataLength <= 0) {
+            LogErr(VB_SEQUENCE, "VariableHeader has 0 length data: %c%c", header[readIndex], header[readIndex + 1]);
+            
+            // empty data, advance only the length of the 2 byte code
+            readIndex += VariableCodeSize;
+        } else if (readIndex + VariableCodeSize + dataLength > header.size()) {
+            // ensure the data length is contained within the header
+            // this is primarily protection against hand modified, or corrupted, sequence files
+            LogErr(VB_SEQUENCE, "VariableHeader '%c%c' has out of bounds data length: %d bytes, max length: %d bytes", header[readIndex], header[readIndex + 1], readIndex + VariableCodeSize + dataLength, header.size());
+            
+            // there is no reasonable way to recover from this error - the reported dataLength is longer than possible
+            // return from parsing variable headers and let the program attempt to read the rest of the file
+            return;
+        } else if (!isRecongizedVariableHeader(header[readIndex], header[readIndex + 1])) {
+            // avoid reading unrecongized variable headers
+            // each 2 byte length can consume up to 65k of memory when allocating vheader->data
+            LogErr(VB_SEQUENCE, "Unknown VariableHeader code: %c%c, length: %d bytes", header[readIndex], header[readIndex + 1], dataLength);
+            
+            // advance the length of the 2 byte code + the length of the data
+            readIndex += VariableCodeSize + dataLength;
```

Comment:
This is the only part of the change that I'm concerned about.   Having a list of "known"  variable headers and enforcing rules on those (like the null termination) makes a lot of sense.   However, ignoring the non-recognized headers is a bad idea.    For things like FPP Connect that copy the headers over to the new fseq files that it generates, those extra headers would be lost.     On the FPP side, the fsequtils command line utility can be used to convert fseq files from format to format (or compression level/type to another) and again, these headers would be lost.   While I'm not sure if anything is using the variable headers outside the known headers, the idea of losing data seems to be a step in the wrong direction.

Reply:
I wasn't aware it utilized in the sole in memory copy as its source of truth of re-encoding output. I'll move this from hard fail to a warning. :+1: